### PR TITLE
Partially port `cpu` to aarch64

### PR DESCRIPTION
**Pull request by NathanRoyer**
This PR implements CpuId for aarch64. Some other `cpu` functions were temporarily implemented as dummies: I think the multicore ARM switch should happen in another PR.

Three CpuId flavors are implemented:
- CpuId: an opaque type wrapping a u32, the compact representation
- MpidrValue: a wrapper for the value which is read from the MPIDR_EL1 register. The inner u64 can be obtained through an unsafe method; this should be used by PSCI code when we want to bring another core up.
- OptionalCpuId, which is used by `task_struct`:
```rust
/// An equivalent to Option<CpuId>, which internally encodes None as
/// `u32::MAX`, which is an invalid CpuId (bits [4:7] of affinity level
/// 0 must always be cleared). This guarantees that it compiles down to
/// lock-free native atomic instructions when using it inside of an atomic
/// type like [`AtomicCell`], as u32 is atomic when running on ARMv8.
```

The PR also contains two unrelated & unimportant modifications which are:
- ~~the removal of a whitespace character at the end of a line in `sleep`'s manifest~~
- the switch to more compact arch gates in `interrupt`

### CODE REVIEW

File: kernel/cpu/src/aarch64.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    /// Valid affinity levels are 0, 1, 2, 3
+    pub fn affinity(self, level: u8) -> u8 {
+        assert!(level < 4, "Valid affinity levels are 0, 1, 2, 3");
+
+        (self.0 >> (level * 8)) as u8
+    }
+}
+
+impl MpidrValue {
+    pub unsafe fn get(self) -> u64 {
```
Comment (kevinaboos) [inline kernel/cpu/src/aarch64.rs]: what's unsafe about this?

File: kernel/cpu/src/aarch64.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+impl From<OptionalCpuId> for Option<CpuId> {
+    fn from(val: OptionalCpuId) -> Self {
+        match val.0 {
+            u32::MAX => None,
+            v => Some(CpuId(v)),
+        }
+    }
+}
```
Comment (kevinaboos) [inline kernel/cpu/src/aarch64.rs:108]: So, the manual way of doing this won't result in the same performance as the compiler-internal form of NPO (null pointer optimization, which is what this technique is called). Unfortunately the only way that NPO is exposed outside of rustc itself is via the NonZero* or NonNull types.

If we want to support a value of 0 directly, we can pretty easily get around that with some simple arithmetic. See this SO comment for an example of how: https://stackoverflow.com/questions/30414068/can-i-use-the-null-pointer-optimization-for-my-own-non-pointer-types#comment48914781_30414145

Comment (kevinaboos) [reply kernel/cpu/src/aarch64.rs:108]: Really, we should make decisions based on performance. The 3 alternatives are:
* my old alignment-based trick
* your new manual option with u32::MAX as the impossible sentinel value
* an existing NonZeroU32 type with a bit of addition math to adjust the real value.

File: kernel/cpu/src/lib.rs [hunk 3]
```diff
@@ -6,11 +6,8 @@
...
 
-#[cfg(target_arch = "x86_64")]
-pub use apic::{
-    CpuId,
-    cpu_count,
-    bootstrap_cpu,
-    is_bootstrap_cpu,
-    current_cpu,
-};
+#[cfg_attr(target_arch = "x86_64", path = "x86_64.rs")]
```
Comment (kevinaboos) [inline kernel/cpu/src/lib.rs:13]: the docs on Line 5 of this crate need an update.

File: kernel/sleep/Cargo.toml [hunk 4]
```diff
@@ -14,7 +14,7 @@ version = "1.2.0"
 path = "../task"
 
 [dependencies.irq_safety]
-git = "https://github.com/theseus-os/irq_safety" 
+git = "https://github.com/theseus-os/irq_safety"
```
Comment (kevinaboos) [inline kernel/sleep/Cargo.toml]: undo changes to this unrelated file

**Review by kevinaboos — CHANGES_REQUESTED** [review]

File: kernel/cpu/src/aarch64.rs (hunk 1 above)
Comment (NathanRoyer) [reply kernel/cpu/src/aarch64.rs]: I wanted to indicate that should not be used often; I forgot to document that by the way

File: kernel/cpu/src/aarch64.rs (hunk 2 above)
Comment (NathanRoyer) [reply kernel/cpu/src/aarch64.rs:108]: Oh, indeed we can store the cpu id "plus one" so that zero would be available for use by None. This is completely the same as storing None as u32::MAX except it adds an addition/subtraction in the conversion. I think my solution is simple and the doc explains it clearly. Perhaps I should add a showcase/example in the doc?

File: kernel/cpu/src/aarch64.rs (hunk 1 above)
Comment (NathanRoyer) [reply kernel/cpu/src/aarch64.rs]: New doc:
```rust
    /// Obtain the inner raw u64 that was read from the MPIDR_EL1 register
    ///
    /// This is not dangerous, the unsafe modifier is simply there to
    /// indicate that this value should only be used when interacting with
    /// hardware/peripherals, in driver code for instance; it should not be
    /// used to identify/reference a CPU core.
```

Comment (kevinaboos) [reply kernel/cpu/src/aarch64.rs]: ok i see, then that's not the proper usage of unsafe. Getting an inner raw value is canonically safe in all Rust libraries, e.g., obtaining a pointer to something. _Using_ that raw value to re-obtain a higher-typed value is what's unsafe, e.g., dereferencing a pointer into a reference, or in this case, converting a raw integer into a `MpidrValue`.

We don't and shouldn't offer a public-facing way to convert a raw integer into a `MpidrValue`, so no unsafety is needed.

Rather than using unsafe, the way to indicate that a raw value should not be used to identify a CPU core is to _prevent_ it from happening by accepting only a higher-typed value that originated from a getter function, not a raw value.

File: kernel/cpu/src/aarch64.rs (hunk 2 above)
Comment (kevinaboos) [reply kernel/cpu/src/aarch64.rs:108]: I was suggesting that you could try the 3 options and see which is more performant; alternatively we can implement the 3 options and gate them behind cfgs for future testing. Your solution is indeed simple but it has branches, which is usually what kills perf, not addition.

No need to worry about it now I suppose, but we can experiment soon.

File: kernel/cpu/src/aarch64.rs (hunk 1 above)
Comment (NathanRoyer) [reply kernel/cpu/src/aarch64.rs]: > Rather than using unsafe, the way to indicate that a raw value should not be used to identify a CPU core is to _prevent_ it from happening by accepting only a higher-typed value that originated from a getter function, not a raw value.

I agree, however driver crates will need to access that value...

I will remove the unsafe modifier.

Comment (kevinaboos) [reply kernel/cpu/src/aarch64.rs]: > I agree, however driver crates will need to access that value...

Right, and they can do some completely safely when obtaining that value from a bonafide instance of `MpidrValue` or something that was constructed from it, e.g., `CpuId`. This is the underlying principle of how type-carried invariants work, we're not really reinventing the wheel here.

> I will remove the unsafe modifier.

Great, thanks. Resolved.

**Review by kevinaboos — APPROVED** [review]
